### PR TITLE
Add EntityPotionEffectImmunityEvent

### PR DESCRIPTION
**Pull request by kivibot**
Allows plugins to selectively disable vanilla potion effect immunities. 
Can also be used to detect if immune mobs have been hit with those effects.

Somewhat related issue #3580

### CODE REVIEW

File: Spigot-Server-Patches/0619-Add-EntityPotionEffectImmunityEvent.patch [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+         } else {
+             MobEffect mobeffect1 = (MobEffect) this.effects.get(mobeffect.getMobEffect());
+@@ -941,6 +943,21 @@ public abstract class EntityLiving extends Entity {
+             }
+         }
+     }
++	
++    // Paper start
++    private boolean checkEffectImmunity(MobEffect mobEffect, EntityPotionEffectEvent.Cause cause) {
++        if(this.d(mobEffect)) {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0619-Add-EntityPotionEffectImmunityEvent.patch]: Obfhelper

File: Spigot-Server-Patches/0619-Add-EntityPotionEffectImmunityEvent.patch [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
++import org.bukkit.craftbukkit.potion.CraftPotionUtil; // Paper
+ 
+ public abstract class EntityLiving extends Entity {
+ 
+@@ -909,7 +911,7 @@ public abstract class EntityLiving extends Entity {
+         }
+         // CraftBukkit end
+ 
+-        if (!this.d(mobeffect)) {
++        if (this.checkEffectImmunity(mobeffect, cause)) { // Paper
```
Comment (Proximyst) [inline Spigot-Server-Patches/0619-Add-EntityPotionEffectImmunityEvent.patch]: Why has the logic been inversed?

**Review by Proximyst — CHANGES_REQUESTED** [review]

**Review by Proximyst — CHANGES_REQUESTED** [review]
See previous comment

**Review by electronicboy — APPROVED** [review]
Provisionally approving, an obfhelper on the actual method would be nice

**Review by electronicboy — APPROVED** [review]

**Comment by Owen1212055** [comment]
This is in general covered by config values, is an event really needed here?

**Comment by Owen1212055** [comment]
Closing, in general, most of this behavior is doable with config options now. 
If someone else still sees use for this and wants to pick up on this feel free. 😄

**Pull request closed without merging**